include/xgboost/intrusive_ptr.h

```diff
@@ -0,0 +1,187 @@
+/*!
+ * Copyright (c) by Contributors 2020
+ * \file intrusive_ptr.h
+ * \brief Implementation of Intrusive Ptr.
+ */
+#ifndef XGBOOST_INTRUSIVE_PTR_H_
+#define XGBOOST_INTRUSIVE_PTR_H_
+
+#include <atomic>
+#include <cinttypes>
+#include <functional>
+
+namespace xgboost {
+/*!
+ * \brief Helper class for embedding reference counting into client objects.
```

Comment:
We should document why it's okay to use `std:: memory_order_relaxed` when incrementing the reference count.
```suggestion
 * \brief Helper class for embedding reference counting into client objects.
 * See https://www.boost.org/doc/libs/1_74_0/doc/html/atomic/usage_examples.html
 * for discussions of memory order.
```